### PR TITLE
[MM-56088] Adding extra logs for openId state failures

### PR DESCRIPTION
**Pull request by BenCookie95**
#### Summary
I was debugging an issue with O365 OpenID login and the invalid state error is annoying because you can't pinpoint exactly which line of code is causing the issue. We return the same error multiple times in order to obfuscate the real error from the frontend.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-56088

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: server/channels/app/oauth.go [hunk 1]
```diff
@@ -822,16 +822,18 @@ func (a *App) AuthorizeOAuthUser(c request.CTX, w http.ResponseWriter, r *http.R
...
 	stateAction := stateProps["action"]
 	if stateAction == model.OAuthActionEmailToSSO && stateEmail == "" {
+		c.Logger().Error("No email provided in state when trying to switch from email to SSO")
 		return nil, "", stateProps, nil, model.NewAppError("AuthorizeOAuthUser", "api.user.authorize_oauth_user.invalid_state.app_error", nil, "", http.StatusBadRequest)
 	}
 
 	cookie, cookieErr := r.Cookie(CookieOAuth)
 	if cookieErr != nil {
-		return nil, "", stateProps, nil, model.NewAppError("AuthorizeOAuthUser", "api.user.authorize_oauth_user.invalid_state.app_error", nil, "", http.StatusBadRequest)
+		return nil, "", stateProps, nil, model.NewAppError("AuthorizeOAuthUser", "api.user.authorize_oauth_user.invalid_state.app_error", nil, "", http.StatusBadRequest).Wrap(cookieErr)
```
Comment (hanzei) [inline server/channels/app/oauth.go:831]: Does the wrapper error get sent to the client? If so, is that an issue?

File: server/channels/app/oauth.go [hunk 2]
```diff
@@ -822,16 +822,18 @@ func (a *App) AuthorizeOAuthUser(c request.CTX, w http.ResponseWriter, r *http.R
 	stateEmail := stateProps["email"]
 	stateAction := stateProps["action"]
 	if stateAction == model.OAuthActionEmailToSSO && stateEmail == "" {
+		c.Logger().Error("No email provided in state when trying to switch from email to SSO")
```
Comment (hanzei) [inline server/channels/app/oauth.go:825]: Is that really an `Error`?

File: server/channels/app/oauth.go [hunk 3]
```diff
@@ -822,16 +822,18 @@ func (a *App) AuthorizeOAuthUser(c request.CTX, w http.ResponseWriter, r *http.R
...
 
 	cookie, cookieErr := r.Cookie(CookieOAuth)
 	if cookieErr != nil {
-		return nil, "", stateProps, nil, model.NewAppError("AuthorizeOAuthUser", "api.user.authorize_oauth_user.invalid_state.app_error", nil, "", http.StatusBadRequest)
+		return nil, "", stateProps, nil, model.NewAppError("AuthorizeOAuthUser", "api.user.authorize_oauth_user.invalid_state.app_error", nil, "", http.StatusBadRequest).Wrap(cookieErr)
 	}
 
 	expectedTokenExtra := generateOAuthStateTokenExtra(stateEmail, stateAction, cookie.Value)
 	if expectedTokenExtra != expectedToken.Extra {
+		c.Logger().Error("Extra token value does not match token generated from state")
```
Comment (hanzei) [inline server/channels/app/oauth.go:836]: Same question regarding `Error` here.

**Review by hanzei — CHANGES_REQUESTED** [review]
It's a bummer that the app layer tries to obfuscate errors. Does it make sense to do something similar to the login code (https://github.com/mattermost/mattermost/blob/ef3834b17bc4b03d20e1ba70c5a9d21b0c68fd6d/server/channels/api4/user.go#L1838-L1857) and obfuscate errors on the API layer, so the logs still contain meaningful errors?

File: server/channels/app/oauth.go (hunk 1 above)
Comment (JulienTant) [reply server/channels/app/oauth.go:831]: > Does the wrapper error get sent to the client? If so, is that an issue?

I reckon only `T('api.user.authorize_oauth_user.invalid_state.app_error')` is returned, but the [r.Cookie() function only returns a `http.ErrNoCookie` ](https://cs.opensource.google/go/go/+/refs/tags/go1.21.5:src/net/http/request.go;l=426) so nothing bad would be returned there.

**Review by JulienTant — APPROVED** [review]
LGTM

File: server/channels/app/oauth.go (hunk 1 above)
Comment (BenCookie95) [reply server/channels/app/oauth.go:831]: All that gets returned to the client is `api.user.authorize_oauth_user.invalid_state.app_error`. Adding the `Wrap` just gives us extra info in the logs

File: server/channels/app/oauth.go [hunk 4]
```diff
@@ -822,16 +822,18 @@ func (a *App) AuthorizeOAuthUser(c request.CTX, w http.ResponseWriter, r *http.R
 	stateEmail := stateProps["email"]
 	stateAction := stateProps["action"]
 	if stateAction == model.OAuthActionEmailToSSO && stateEmail == "" {
+		c.Logger().Error("No email provided in state when trying to switch from email to SSO")
 		return nil, "", stateProps, nil, model.NewAppError("AuthorizeOAuthUser", "api.user.authorize_oauth_user.invalid_state.app_error", nil, "", http.StatusBadRequest)
```
Comment (hanzei) [inline server/channels/app/oauth.go:826]: I wonder if it's even better to wrap a new error and return it:

```suggestion
		err := errors.New("No email provided in state when trying to switch from email to SSO")
		return nil, "", stateProps, nil, model.NewAppError("AuthorizeOAuthUser", "api.user.authorize_oauth_user.invalid_state.app_error", nil, "", http.StatusBadRequest).Wrap(err)
```

That way a single line the logs contains all the context on what went wrong.

Comment (BenCookie95) [reply server/channels/app/oauth.go]: Yeah I like it, done

**Comment by hanzei** [comment]
@BenCookie95 Can we merge the PR?

**Pull request merged**